### ISSUE TITLE
Remove Catlab dependency

### ISSUE THREAD

**Issue by ChrisRackauckas** [SciML/Catalyst.jl#293]
This is pretty crucial. That library is very unmaintained and has been the root of our CI problems for months now because Catalyst is part of DiffEqProblemLibrary.

**Comment by isaacsas** [comment]
I can try to just copy over the definitions / code we need from Catlab's graph submodule. I think it is just basically one file or so. If it keeps improving and being actively maintained we could ultimately switch to 

https://github.com/JuliaGraphs/GraphPlot.jl

which would have the advantage of working cross-platform out of the box (since the jll for Graphviz doesn't currently work on all platforms we can't require it).

**Comment by ChrisRackauckas** [comment]
@mohamed82008 found that Catlab's type-piracy is the cause of a ton of other issues:

https://github.com/JuliaLang/julia/issues/39534#issuecomment-774439770

including type-piracy in getproperty. I don't think that is the kind of thing we want to be shipping with because it can cause a whole bunch of other problems.

**Comment by ChrisRackauckas** [comment]
Also, Catlab doesn't seem installable on Windows anymore. I cannot use Catalyst without removing Catlab.

```julia
ERROR: LoadError: ArgumentError: Package Indexing [313cdc1a-70c2-5d6a-ae34-0150d3930a38] is required but does not seem to be installed:
 - Run `Pkg.instantiate()` to install all recorded dependencies.
Stacktrace:
 [1] _require(::Base.PkgId) at .\loading.jl:999
 [2] require(::Base.PkgId) at .\loading.jl:928
 [3] require(::Module, ::Symbol) at .\loading.jl:923
 [4] include(::Function, ::Module, ::String) at .\Base.jl:380
 [5] include(::Module, ::String) at .\Base.jl:368
 [6] top-level scope at none:2
 [7] eval at .\boot.jl:331 [inlined]
 [8] eval(::Expr) at .\client.jl:467
 [9] top-level scope at .\none:3
in expression starting at C:\Users\accou\.julia\packages\SplitApplyCombine\RS5bI\src\SplitApplyCombine.jl:4
ERROR: LoadError: Failed to precompile SplitApplyCombine [03a91e81-4c3e-53e1-a0a4-9c0c8f19dd66] to C:\Users\accou\.julia\compiled\v1.5\SplitApplyCombine\vbX6o_6hXNi.ji.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at .\loading.jl:1305
 [3] _require(::Base.PkgId) at .\loading.jl:1030
 [4] require(::Base.PkgId) at .\loading.jl:928
 [5] require(::Module, ::Symbol) at .\loading.jl:923
 [6] include(::Function, ::Module, ::String) at .\Base.jl:380
 [7] include(::Module, ::String) at .\Base.jl:368
 [8] top-level scope at none:2
 [9] eval at .\boot.jl:331 [inlined]
 [10] eval(::Expr) at .\client.jl:467
 [11] top-level scope at .\none:3
in expression starting at C:\Users\accou\.julia\packages\TypedTables\dEa2h\src\TypedTables.jl:5
ERROR: LoadError: LoadError: Failed to precompile TypedTables [9d95f2ec-7b3d-5a63-8d20-e2491e220bb9] to C:\Users\accou\.julia\compiled\v1.5\TypedTables\NU69s_6hXNi.ji.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at .\loading.jl:1305
 [3] _require(::Base.PkgId) at .\loading.jl:1030
 [4] require(::Base.PkgId) at .\loading.jl:928
 [5] require(::Module, ::Symbol) at .\loading.jl:923
 [6] include(::Function, ::Module, ::String) at .\Base.jl:380
 [7] include at .\Base.jl:368 [inlined]
 [8] include(::String) at C:\Users\accou\.julia\packages\Catlab\yfjLt\src\Catlab.jl:1
 [9] top-level scope at C:\Users\accou\.julia\packages\Catlab\yfjLt\src\Catlab.jl:6
 [10] include(::Function, ::Module, ::String) at .\Base.jl:380
 [11] include(::Module, ::String) at .\Base.jl:368
 [12] top-level scope at none:2
 [13] eval at .\boot.jl:331 [inlined]
 [14] eval(::Expr) at .\client.jl:467
 [15] top-level scope at .\none:3
in expression starting at C:\Users\accou\.julia\packages\Catlab\yfjLt\src\categorical_algebra\CSetDataStructures.jl:14
in expression starting at C:\Users\accou\.julia\packages\Catlab\yfjLt\src\Catlab.jl:6
ERROR: LoadError: Failed to precompile Catlab [134e5e36-593f-5add-ad60-77f754baafbe] to C:\Users\accou\.julia\compiled\v1.5\Catlab\fBQ1G_6hXNi.ji.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at .\loading.jl:1305
 [3] _require(::Base.PkgId) at .\loading.jl:1030
 [4] require(::Base.PkgId) at .\loading.jl:928
 [5] require(::Module, ::Symbol) at .\loading.jl:923
 [6] include(::Function, ::Module, ::String) at .\Base.jl:380
 [7] include(::Module, ::String) at .\Base.jl:368
 [8] top-level scope at none:2
 [9] eval at .\boot.jl:331 [inlined]
 [10] eval(::Expr) at .\client.jl:467
 [11] top-level scope at .\none:3
in expression starting at C:\Users\accou\.julia\packages\Catalyst\UvrdM\src\Catalyst.jl:29
ERROR: LoadError: Failed to precompile Catalyst [479239e8-5488-4da2-87a7-35f2df7eef83] to C:\Users\accou\.julia\compiled\v1.5\Catalyst\QGu8a_6hXNi.ji.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at .\loading.jl:1305
 [3] _require(::Base.PkgId) at .\loading.jl:1030
 [4] require(::Base.PkgId) at .\loading.jl:928
 [5] require(::Module, ::Symbol) at .\loading.jl:923
 [6] include(::Function, ::Module, ::String) at .\Base.jl:380
 [7] include(::Module, ::String) at .\Base.jl:368
 [8] top-level scope at none:2
 [9] eval at .\boot.jl:331 [inlined]
 [10] eval(::Expr) at .\client.jl:467
 [11] top-level scope at .\none:3
```

**Comment by isaacsas** [comment]
I can make a point release that drops Catlab and graph functionality quick today, and then we can add it back once it’s ported over...

**Comment by ChrisRackauckas** [comment]
That would be helpful. I was going to look into doing something similar since this all resurfaced last night.

**Comment by isaacsas** [comment]
I’ll take a look in a few minutes. Just need my coffee to finish brewing!

**Comment by isaacsas** [comment]
Closed by https://github.com/SciML/Catalyst.jl/pull/299

**Comment by epatters** [comment]
> This is pretty crucial. That library is very unmaintained...

Just seeing this issue. You are of course free to drop the dependency and rip out the Graphviz code, but your statement that the "library is very unmaintained" is false. It is actively maintained, as a glance at the commit log would show. Whether it meets your standards in other regards is for you to decide.

I'm bummed to learn about the type piracy issue—I didn't realize that what we were doing counted as type piracy—and I'll look into fixing it.

**Comment by isaacsas** [comment]
@epatters I agree; it is clear you all are actively working on Catlab! I imagine Chris was just frustrated dealing with CI failures.

What might be a good way forward on our end, and I think would be really cool, would be if we put together a conversion from Catalyst models to a Catlab-compatible form (maybe through AlgebraicPetri?). We could add a Catalyst transformation to such a form, and then directly leverage all the great functionality you've got in Catlab (including the graph components). To keep the dependencies here slim I've been thinking we might want to have another package for analysis of Catalyst models, with Catalyst.jl just exporting the base API for working with `ReactionSystem`s. (For example, I'd like to add some tooling for detecting conservation laws and transformation to matrix-vector system forms.) 

That would be a way to keep the dependencies down, which would allow Catalyst to be used in critical infrastructure for SciML, while still letting us connect up with other packages.

**Comment by ChrisRackauckas** [comment]
It's not CI: CI is the only place I can run it these days. I haven't been able to do `using Catalyst` for 3 weeks because of compile time error from Catlab. One week in I opened this issue, since erroring upon `using` for a week is pretty insane. By 3 weeks and getting reports from everyone around me that Catalyst is causing issues because of Catlab, it's gotta go. Sorry, but type piracy, upper bounds on a bunch of core packages like Reexport, issues on Windows, and bad interactions with precompilation (which seems to cause some issues with sysimage building) which persist for months is a very potent combination. I was starting to get daily messages about fixing it (because Catalyst is throughout the SciML tests, Pumas, etc.), so it was a bit too much just for a plotting feature. I'm sure it'll grow into something great, but at least right now there seem to be a lot of issues that have a turnaround time on the order of months that are deep enough to be release blocking products and forcing students to choose other projects, so I had to give up on it.

**Comment by isaacsas** [comment]
Yeah, I've had no issues, but I'm on a Mac. In the future, maybe encourage those users to open issues here too, or let me know if you are getting so many reports of problems. I would've made the change when you opened this issue originally if I had realized the depth of the problem... The last few weeks have been busy with the start of classes, so it got back-burnered...

**Comment by isaacsas** [comment]
Maybe we should add back in Windows CI?

**Comment by jpfairbanks** [comment]
I am also hurt by the accusation that Catlab is not maintained when we have tagged 6 releases in the last ~2 months. The AlgebraicJulia team is available on several communication platforms and can help in fixing issues. We use CI for Catlab and our packages that depend on it including AlgebraicPetri with minimal issues. When Chris asked about a problem installing on windows, we tried and were unable to reproduce the failure. I apologize for being unable to assist further but without the ability to reproduce, I am not sure how to help. Since I had not heard back about the issue, I thought it was transient. 

We definitely need to work on compile times and precompilation, but Catlab is a fairly substantial endeavor exploring new approaches to some fundamental aspects of data structures so this hasn't been a large priority yet. Since you are only using Catlab for plotting and not the broader approach, I agree it would be best for you to write your own plotting library. If that includes a stand alone GraphViz drawing tool, that would be generally useful to the community.
